lean4-infoview/src/infoview/messages.tsx

```diff
@@ -118,41 +122,42 @@ export function AllMessages({uri: uri0}: { uri: DocumentUri }) {
  * Provides a `DiagnosticsContext` by subscribing to server diagnostic notifications
  * and querying for interactive diagnostics whenever (LSP standard) diagnostics arrive.
  */
-export function WithDiagnosticsContext(props: React.PropsWithChildren<any>) {
-    const rs = React.useContext(RpcContext)
-    const sv = React.useContext(VersionContext)
+export function WithLspDiagnosticsContext({children}: React.PropsWithChildren<{}>) {
     const [allDiags, _0] = useServerNotificationState(
         'textDocument/publishDiagnostics',
-        new Map<DocumentUri, InteractiveDiagnostic[]>(),
-        async (params: PublishDiagnosticsParams) => {
-            let diags: InteractiveDiagnostic[] | undefined
-
-            if (!sv.hasWidgetsV1()) {
-                diags = params.diagnostics.map(d => { return { ...(d as LeanDiagnostic), message: { text: d.message } } })
-            } else try {
-                diags = await getInteractiveDiagnostics(rs, { uri: params.uri, line: 0, character: 0 })
-            } catch (err: any) {}
-            if (diags)
-                return allDiags => {
-                    const newMap = new Map(allDiags)
-                    return newMap.set(params.uri, diags!)
-                }
-            else return allDiags => allDiags
-        },
+        new Map<DocumentUri, Diagnostic[]>(),
+        async (params: PublishDiagnosticsParams) => allDiags =>
+            new Map(allDiags).set(params.uri, params.diagnostics),
         []
     )
 
-    return (
-        <DiagnosticsContext.Provider value={allDiags}>
-            {props.children}
-        </DiagnosticsContext.Provider>
-    );
+    return <LspDiagnosticsContext.Provider value={allDiags}>{children}</LspDiagnosticsContext.Provider>
+}
+
+export function useMessagesForFile(uri: DocumentUri, line?: number): InteractiveDiagnostic[] {
+    const rs = React.useContext(RpcContext)
+    const sv = React.useContext(VersionContext)
+    const lspDiags = React.useContext(LspDiagnosticsContext)
+    const [diags, setDiags] = React.useState<InteractiveDiagnostic[]>([])
+    async function updateDiags() {
+        setDiags((lspDiags.get(uri) || []).map(d => ({ ...(d as LeanDiagnostic), message: { text: d.message } })));
+        if (sv?.hasWidgetsV1()) {
+            try {
+                const diags = await getInteractiveDiagnostics(rs, { uri, line: 0, character: 0 },
+                    line ? { start: line, end: line + 1 } : undefined)
+                if (diags) {
+                    setDiags(diags)
+                }
+            } catch (err: any) {
+                console.log('getInteractiveDiagnostics error ', err)
+            }
+        }
```

Comment:
This breaks backwards compatibility with pre-line-filterable-diags servers, no? Since we no longer make a request if `!hasWidgetsV1()`.